src/network/protocol/messages/version.zig

```diff
@@ -2,200 +2,224 @@ const std = @import("std");
 const native_endian = @import("builtin").target.cpu.arch.endian();
 const protocol = @import("../lib.zig");
 
-const NetworkAddress = protocol.NetworkAddress;
 const ServiceFlags = protocol.ServiceFlags;
-const IpV6Address = protocol.IpV6Address;
-const Message = protocol.message.Message;
 
 const Endian = std.builtin.Endian;
+const Sha256 = std.crypto.hash.sha2.Sha256;
 
 const CompactSizeUint = @import("bitcoin-primitives").types.CompatSizeUint;
 
 /// VersionMessage represents the "version" message
 ///
 /// https://developer.bitcoin.org/reference/p2p_networking.html#version
 pub const VersionMessage = struct {
-    version: i32,
-    services: u64,
+    recv_ip: [16]u8,
+    trans_ip: [16]u8,
     timestamp: i64,
-    addr_recv: NetworkAddress,
-    addr_trans: NetworkAddress,
+    services: u64,
     nonce: u64,
-    user_agent: ?[]const u8,
+    recv_services: u64,
+    trans_services: u64,
+    version: i32,
     start_height: i32,
+    recv_port: u16,
+    trans_port: u16,
+    user_agent: ?[]const u8,
     relay: ?bool,
 
-    /// Will free the user_agent if present in the message
+    pub fn name() [12]u8 {
+        var ret = std.mem.zeroes([12]u8);
+        @memcpy(ret[0..7], protocol.CommandNames.VERSION);
+        return ret;
+    }
+
+    /// Returns the message checksum
+    ///
+    /// Computed as `Sha256(Sha256(self.serialize()))[0..4]`
+    pub fn checksum(self: VersionMessage) [4]u8 {
+        var digest: [32]u8 = undefined;
+        var hasher = Sha256.init(.{});
+        const writer = hasher.writer();
+        self.serializeToWriter(writer) catch unreachable; // Sha256.write is infaible
+        hasher.final(&digest);
+
+        Sha256.hash(&digest, &digest, .{});
```

Comment:
i am not sure, but in bitcoin-primitives should double sha256

Reply:
I looked for an exposed method in bitcoin-primitives doing that but I couldn't find it